### DISCUSSION TITLE
TimeEvent only for calculation process

### DISCUSSION THREAD

**Discussion by aLAN-LDZ** [HomeSpan/HomeSpan#187] [Q&A] [answered]
Hi

Currently part of my code for calculating data and uploading it to the http server (for create charts) is performed in one of the sensor blocks.

It's not very pretty. Hence my question, how can I make a block only for these calculations so that it executes once every 30s? 

And can I put it in the main file right under: homeSpan.poll();?

**Comment by SnowballsChance** [comment, marked as answer]
For things like this I have triggered it  in `void loop()` and held the data as a global.

````C++
// Main loop non-blocking timers
#define INTERVAL1 100000
#define INTERVAL2 30000
unsigned long time_1 = 0;
unsigned long time_2 = 0;
`

....snip....


````C++
void loop() {
  do { //infinite loop
      // code here is processed every cycle    
    
    homeSpan.poll();
 
    if (millis() - time_1 >= INTERVAL1) { 

      //Serial.println("Interval 1");
      // Insert code here

      time_1 = millis();
    } // end interval 1

    if (millis() - time_2 >= INTERVAL2) {

      // Serial.println("Interval 2");

      // Insert code here

        go_and_do_stuff();
     
      time_2 = millis();
    } // end interval 2

  } while (1); //infinite loop

}   // end loop()
```

Hope his helps

**Comment by aLAN-LDZ** [comment]
thank you for the hint, isn't it that the millis () function starts counting again after some time?

definition:

`Returns the number of milliseconds passed since the Arduino board began running the current program. This number will overflow (go back to zero), after approximately 50 days.`

What in this case?

**Comment by SnowballsChance** [comment]
you are welcome.

Yes that is right after about 50 days as mentioned. The way I have implemented it using an unsigned long and subtracting the interval time the math works out even when millis() rolls over. 

This post [here](https://www.norwegiancreations.com/2018/10/arduino-tutorial-avoiding-the-overflow-issue-when-using-millis-and-micros/) gives a good explanation. 

Cheers
